### PR TITLE
stub: Document StreamObserver is an async API

### PR DESCRIPTION
**Pull request by ejona86**
Missing docs were brought up in #8423

CC @morgwai

### CODE REVIEW

File: stub/src/main/java/io/grpc/stub/ClientCallStreamObserver.java [hunk 1]
```diff
@@ -20,7 +20,8 @@
 
 /**
  * A refinement of {@link CallStreamObserver} that allows for lower-level interaction with
- * client calls.
+ * client calls. An instance of this class is obtained via {@link ClientResponseObserver}, or by
```
Comment (dapengzhang0) [inline stub/src/main/java/io/grpc/stub/ClientCallStreamObserver.java:23]: `{@link ClientResponseObserver#beforeStart()"` to be more precisely?

**Comment by ejona86** [comment]
@dapengzhang0, friendly ping

File: stub/src/main/java/io/grpc/stub/StreamObserver.java [hunk 2]
```diff
@@ -31,6 +31,16 @@
  * not need to be synchronized together; incoming and outgoing directions are independent.
  * Since individual {@code StreamObserver}s are not thread-safe, if multiple threads will be
  * writing to a {@code StreamObserver} concurrently, the application must synchronize calls.
+ *
+ * <p>This API is asynchronous, so methods may return before the operation completes. The API
+ * provides no guarantees for how quickly an operation will complete, so utilizing flow control via
+ * {@link ClientCallStreamObserver} and {@link ServerCallStreamObserver} to avoid excessive
+ * buffering is recommended for streaming RPCs. gRPC's implementation of {@code onError()} on
+ * client-side causes the RPC to be cancelled and discards all messages, so completes quickly.
```
Comment (dapengzhang0) [inline stub/src/main/java/io/grpc/stub/StreamObserver.java:39]: "so completes quickly" - How quickly is quick? Does this behavior guarantee anything users can rely on?

File: stub/src/main/java/io/grpc/stub/ClientCallStreamObserver.java (hunk 1 above)
Comment (ejona86) [reply stub/src/main/java/io/grpc/stub/ClientCallStreamObserver.java:23]: Does that verbosity actually help? There's only one method...

File: stub/src/main/java/io/grpc/stub/StreamObserver.java (hunk 2 above)
Comment (ejona86) [reply stub/src/main/java/io/grpc/stub/StreamObserver.java:39]: No, we aren't going to define how quickly. I think it is helpful to understand the system and what things do. And it is also useful to state this behavior applies only to client-side. If a user saw client-side cleaned up quickly, this documentation would define that behavior and tell them to not assume server-side is the same.

**Review by dapengzhang0 — APPROVED** [review]
LGTM

**Pull request merged**